pallets/ethereum-transactions/src/tests/tests.rs

```diff
@@ -0,0 +1,1287 @@
+        response,
+        headers: vec![],
+        body: body.encode(),
+        sent: true,
+        ..Default::default()
+    });
+}
+
+fn mark_avn_tx_as_sent(tx_id: TransactionId, eth_tx_hash: H256) {
+    assert_ok!(<EthereumTransactions as Store>::Repository::mutate(tx_id, |tx| {
+        tx.set_eth_tx_hash::<TestRuntime>(eth_tx_hash)
+    }));
+}
+
+#[test]
+fn validate_generated_publish_root_bytes_abi_success() {
+    let mut ext = ExtBuilder::build_default()
+        .with_genesis_config()
+        .with_validators()
+        .as_externality();
+
+    ext.execute_with(|| {
+        let (_, _, _, candidate_tx) = generate_merkle_root_mock_data_with_signatures();
+        let tx = candidate_tx.to_abi(EthereumTransactions::get_publish_root_contract()).unwrap();
+
+        assert_eq!(hex::encode(tx.data), ENCODED_DATA);
+    });
+}
+
+// ========================== Test Module functions: public interface
+// ===============================
+
+// TODO: rename to TestSetupBuilder
+// or move static methods to Mocks
+struct DefaultTransactionBuilder {
+    tx_type: EthTransactionType,
+    sender: Option<AccountId>,
+    confirmations: Vec<ecdsa::Signature>,
+}
+
+impl DefaultTransactionBuilder {
+    fn new(tx_type: EthTransactionType) -> Self {
+        Self { tx_type, sender: None, confirmations: vec![] }
+    }
+
+    pub fn default_roothash() -> [u8; 32] {
+        return [3u8; 32]
+    }
+
+    // The quorum is 2/3 of the validators plus 1
+    // Our validator list has 3 members, so the quorum must be 3
+    // I chose not to implement the same logic of the quorum function
+    // because that would defeat the point of the test being black box
+    pub fn default_quorum() -> u32 {
+        assert_eq!(mock::AVN::active_validators().len(), 3);
+        3
+    }
+
+    pub fn build_default() -> Self {
+        let tx_type = EthTransactionType::PublishRoot(PublishRootData {
+            root_hash: Self::default_roothash(),
+        });
+
+        Self::new(tx_type)
+    }
+
+    pub fn with_sender(&mut self, sender: AccountId) -> &mut Self {
+        self.sender = Some(sender);
+        self
+    }
+
+    pub fn with_confirmations(&mut self) -> &mut Self {
+        self.confirmations.append(&mut vec![
+            ecdsa::Signature::from_raw([1; 65]),
+            ecdsa::Signature::from_raw([2; 65]),
+            ecdsa::Signature::from_raw([3; 65]),
+        ]);
+        self
+    }
+
+    pub fn as_candidate(&self, tx_id: TransactionId) -> EthTransactionCandidate {
+        // Returns an EthTransactionCandidate that should be equal to what the Module produces when
+        // submitting the tx_type The module obtains the tx_id by reading the nonce by
+        // calling get_unique_transaction_identifier() If we call this here, then the module
+        // will always generate a different tx_id Therefore, this function does not use the
+        // same method, to avoid modifying the test state
```

Comment:
```suggestion
        // Returns an EthTransactionCandidate that should be equal to what the Module produces when
        // submitting the tx_type. The module obtains the tx_id by reading the nonce by
        // calling get_unique_transaction_identifier(). If we call this here, then the module
        // will always generate a different tx_id. Therefore, this function does not use the
        // same method, to avoid modifying the test state.
```